### ISSUE TITLE
Avoid type erasure from the mock vault

### ISSUE THREAD

**Issue by lionel1704** [maidsafe/safe_client_libs#884]
Currently, the `DataId` enum in the mock vault has only Mutable, AppendOnly and Immutable types. As mentioned in #883, published and unpublished variants will not reuse the same address space. To facilitate this and to also avoid type erasure the DataId enum should be:
```
enum DataId {
    PubImmutable(address),
    UnpubImmutable(address),
    UnPubMutable(address, tag),
    PubAppendOnly(address, tag),
    UnpubAppendOnly(address, tag),
    ...
}
```
The `Data` enum in the mock vault should also follow the same pattern.

**Comment by m-cat** [comment]
Instead of storing the name and the tag, we could just store the address types (e.g. `MDataAddress`), which contain the name and tag as well as the data type (Seq, Unseq, etc.). Since the address types encode the type already, this would avoid type erasure (where we were storing the name and tag but not the type of idata, mdata, or adata it came from). So we don't really need all these variants for `DataID`, it may be more explicit than necessary and we could be utilizing the existing types instead:

```rust
enum DataId {
    Immutable(IDataAddress),
    Mutable(MDataAddress),
    AppendOnly(ADataAddress),
}
```

**Comment by lionel1704** [comment]
Resolved in #886
